### PR TITLE
Copy edge boundary conditions in all_first_order()

### PR DESCRIPTION
**Pull request by roystgnr**
Previously we were only preserving node and side conditions.

This actually turned out to be a red herring w.r.t. the bug I was
hunting, but it's nice to fix it too.

### CODE REVIEW

**Comment by jwpeterson** [comment]
:+1:

**Comment by dknez** [comment]
So now we need to update this PR to copy shellfaces too, right?

**Comment by roystgnr** [comment]
Right, but honestly I'm wondering if we need to rethink how we handle boundary conditions.  Those "raw" boundary condition calls were a hack and now it's starting to show.  Something like a BoundaryInfo::copy_boundary_conditions(old_elem, new_elem) would be a better fit for here, for example, and then when adding new boundary condition types we'd only have to alter BoundaryInfo rather than finding every BoundaryInfo::raw* user.

Perhaps a serialization and unserialization function as well?  One thing I missed when looking over #948 was parallel_elem.C - there's a good chance that DistributedMesh and even ReplicatedMesh will lose shellface BCs because we're not copying them with the others.

**Comment by dknez** [comment]
> Right, but honestly I'm wondering if we need to rethink how we handle boundary conditions.  Those "raw" boundary condition calls were a hack and now it's starting to show.  Something like a BoundaryInfo::copy_boundary_conditions(old_elem, new_elem) would be a better fit for here, for example, and then when adding new boundary condition types we'd only have to alter BoundaryInfo rather than finding every BoundaryInfo::raw* user.

Yeah, I agree. The current implementation could use some cleaning up. BoundaryInfo::copy_boundary_conditions sounds like a good idea to me.

**Comment by roystgnr** [comment]
Rebased and refactored.  Hopefully this is good to merge now.  I'll put the parallel_elem.C fix into another PR.

**Comment by dknez** [comment]
Looks good to me.

Beyond this PR, are there any other updates/changes to BoundaryInfo that you have in mind?

**Comment by roystgnr** [comment]
Depends; do you have any other changes in mind?  ;-)

If BoundaryInfo is going to remain stable for a couple years then I'll just add the shellface (un)serialization directly to ParallelElem; if not then I'll want to factor that out too.

**Comment by dknez** [comment]
I don't have any other changes in mind :) The shellface stuff was all I needed.

File: include/mesh/boundary_info.h [hunk 1]
```diff
@@ -504,6 +504,15 @@ class BoundaryInfo : public ParallelObject
                          const unsigned short int side,
                          std::vector<boundary_id_type> & vec_to_fill) const;
 
+  /*
+   * Copy boundary ids associated with old_elem (but not its nodes)
+   * from old_boundary_info (which may be this) into this boundary
+   * info, associating them with new_elem.
+   */
+  void copy_boundary_ids (const BoundaryInfo &old_boundary_info,
```
Comment (jwpeterson) [inline include/mesh/boundary_info.h]: Need space on both sides of ampersand.

File: src/mesh/boundary_info.C [hunk 2]
```diff
@@ -1285,6 +1285,37 @@ void BoundaryInfo::raw_boundary_ids (const Elem * const elem,
 
 
 
+void BoundaryInfo::copy_boundary_ids
+  (const BoundaryInfo &old_boundary_info,
```
Comment (jwpeterson) [inline src/mesh/boundary_info.C]: Space on both sides of ampersand, also this will be reindented by the reindent script.

**Pull request merged**